Spoiler.py

```diff
@@ -285,3 +288,89 @@ def create_playthrough(self) -> None:
 
         if worlds[0].entrance_shuffle:
             self.entrance_playthrough = OrderedDict((str(i + 1), list(sphere)) for i, sphere in enumerate(entrance_spheres))
+
+
+class Copier:
+    def __init__(self, spoiler: Spoiler) -> None:
+        self.spoiler: Spoiler = spoiler
+        self.worlds: dict[int, World] = {}
+        self.dungeons: dict[int, Dungeon] = {}
+        self.regions: dict[int, Region] = {}
+        self.entrances: dict[int, Entrance] = {}
+        self.locations: dict[int, Location] = {}
+        self.items: dict[int, Item] = {}
+
+    def copy(self) -> list[World]:
+        if self.worlds:
+            return list(self.worlds.values())
+
+        # Make copies.
+        for world in self.spoiler.worlds:
+            self.worlds[id(world)] = world.copy()
+            for dungeon in world.dungeons:
+                self.dungeons[id(dungeon)] = dungeon.copy()
+                for item in chain(dungeon.boss_key, dungeon.small_keys, dungeon.dungeon_items, dungeon.silver_rupees):
+                    if id(item) in self.items:
+                        continue
+                    self.items[id(item)] = item.copy()
+            for region in world.regions:
+                self.regions[id(region)] = region.copy()
+                for entrance in chain(region.entrances, region.exits, [region.savewarp] if region.savewarp else []):
+                    if id(entrance) in self.entrances:
+                        continue
+                    self.entrances[id(entrance)] = entrance.copy()
+                for location in region.locations:
+                    if id(location) in self.locations:
+                        continue
+                    self.locations[id(location)] = location.copy()
+                    if location.item and id(location.item) not in self.items:
+                        self.items[id(location.item)] = location.item.copy()
+            for item in world.itempool:
+                if id(item) in self.items:
+                    continue
+                self.items[id(item)] = item.copy()
+
+        # Update references.
+        for world in self.worlds.values():
+            world.dungeons = [self.dungeons.get(id(dungeon), dungeon) for dungeon in world.dungeons]
+            world.regions = [self.regions.get(id(region), region) for region in world.regions]
+            world.itempool = [self.items.get(id(item), item) for item in world.itempool]
+
+        for dungeon in self.dungeons.values():
+            dungeon.world = self.worlds.get(id(dungeon.world), dungeon.world)
```

Comment:
I'm a bit confused why these steps are necessary, shouldn't these fields already be set when the dungeon is copied in on line 311?

Reply:
The initial copy references the original objects, not the copied versions. This is where recursion was used previously - it just made sure that the copy was made so that it could reference the copy. This version instead just makes more shallow copies and then updates the references to point to the copies at the end.